OpenSSL/SSL.py

```diff
@@ -809,6 +812,92 @@ def wrapper(ssl, alert, arg):
         _lib.SSL_CTX_set_tlsext_servername_callback(
             self._context, self._tlsext_servername_callback)
 
+
+    def set_npn_advertise_callback(self, callback):
+        """
+        Specify a callback function that will be called when offering `Next
+        Protocol Negotiation
+        <https://technotes.googlecode.com/git/nextprotoneg.html>`_ as a server.
+
+        :param callback: The callback function.  It will be invoked with one
+            argument, the Connection instance.  It should return a list of
+            bytestrings representing the advertised protocols, like
+            ``[b'http/1.1', b'spdy/2']``.
+        """
+        @wraps(callback)
+        def wrapper(ssl, out, outlen, arg):
+            conn = Connection._reverse_mapping[ssl]
+            protos = callback(conn)
```

Comment:
> What worries me is debuggability. Swallowing exceptions here will make it really tricky for developers to spot their mistakes.

Yep.  I'd definitely like to see something better than just swallowing the exception.

In other places, pyOpenSSL saves the exception and re-raises it at a later point where OpenSSL isn't going to get in the way.  This can make sense in cases where "a later point" is obvious and soon - for example, certificate verification happens in response to a `SSL_read` call (typically).  So it's somewhat straightforward to save verification exceptions and raise them from `Connection.recv` (which is the wrapper for `SSL_read`).

Does NPN fit into a model like that?  Maybe NPN is advertised in response to an `SSL_read` call as well?  After the server receives the `ClientHello`, perhaps?  Take a look at `_VerifyHelper` to see how this is implemented for the certificate verification callback.  I'm not sure any of the code is re-usable (at least, not as currently written) but maybe it will give you some ideas.